helical/models/geneformer/fine_tuning_model.py

```diff
@@ -0,0 +1,203 @@
+            lr_scheduler_params: Optional[dict] = None,
+            silent = False):
+        """Fine-tunes the Geneformer model for classification tasks. 
+
+        Parameters
+        ----------
+
+        train_dataset : Dataset
+            A helical processed dataset for fine-tuning
+        optimizer : torch.optim, default = torch.optim.AdamW
+            The optimizer to be used for training.
+        optimizer_params : dict
+            The optimizer parameters to be used for the optimizer specified. This list should NOT include model parameters.
+            e.g. optimizer_params = {'lr': 0.0001}
+        loss_function : torch.nn.modules.loss, default = torch.nn.modules.loss.CrossEntropyLoss()
+            The loss function to be used.
+        label : str, optional, default = "cell_types"
+            The column in the dataset containing the training labels. These should be stored as unique per class integers.
+        epochs : int, optional, default = 10
+            The number of epochs to train the model
+        freeze_layers : int, optional, default = 0
+            The number of layers to freeze.
+        validation_dataset : Dataset, default = None
+            A helical processed dataset for per epoch validation. If this is not specified, no validation will be performed.
+        lr_scheduler_params : dict, default = None
+            The learning rate scheduler parameters for the transformers get_scheduler method. The optimizer will be taken from the optimizer input and should not be included in the learning scheduler parameters. If not specified, no scheduler will be used.
+            e.g. lr_scheduler_params = { 'name': 'linear', 'num_warmup_steps': 0, 'num_training_steps': 5 }
+
+        """
+                
+        model_input_size = get_model_input_size(self.helical_model)
+
+        cls_present = any("<cls>" in key for key in self.gene_token_dict.keys())
+        eos_present = any("<eos>" in key for key in self.gene_token_dict.keys())
+        if self.emb_mode == "cls":
+            assert cls_present, "<cls> token missing in token dictionary"
+            # Check to make sure that the first token of the filtered input data is cls token
+            cls_token_id = self.gene_token_dict["<cls>"]
+            assert (
+                train_dataset["input_ids"][0][0] == cls_token_id
+            ), "First token is not <cls> token value"
+        elif self.emb_mode == "cell":
+            if cls_present:
+                logger.warning(
+                    "CLS token present in token dictionary, excluding from average."
+                )
+            if eos_present:
+                logger.warning(
+                    "EOS token present in token dictionary, excluding from average."
+                )
+        
+        total_batch_length = len(train_dataset)
+        #initialise optimizer
+        optimizer = optimizer(self.parameters(), **optimizer_params)
+
+        #initialise lr_scheduler
```

Comment:
This function is a bit big. For better overview, do you think you could put initialisations, validations and info logs in another function such as `prepare_training`?

Reply:
Not sure I agree here. I think having all the inits in train makes sense to keep the flow a little bit clearer (insteda of having a lot of nested functrons). I think in a second phase we could use wrappers such as pytorch lightning to make it more readable.

Reply:
For me, I prefer smaller functions because it gives you the opportunity to categorise things under a well chosen name and should thus not hinder the flow. Also it gives you the option to test them one at a time instead of dealing with a monolith. But I'm fine with having it all together :)